### PR TITLE
Inter 06

### PR DESCRIPTION
**Pull request by DavidMistriel**

### CODE REVIEW

File: HeroGame/hero.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import math
+
+
+class Hero:
+    heal_percent = 0.20  # 10%
+    level_up_benefits_percent = 0.50  # 20%
```
Comment (alphabestteam) [inline HeroGame/hero.py:6]: something looks wrong with the comments :(

File: HeroGame/hero.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    level_up_threshold = 3
```
Comment (alphabestteam) [inline HeroGame/hero.py:7]: consts should be upper case

File: HeroGame/hero.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    @defending.setter
+    def defending(self, is_defending: bool) -> None:
+        self._defending = is_defending
+
+    def heal(self) -> None:
+        healing = math.ceil(self.hp * Hero.heal_percent)
+        if healing > self.max_health:
+            self.hp = self.max_health
+        else:
+            self.hp += healing
```
Comment (alphabestteam) [inline HeroGame/hero.py:31]: here you are adding not updating, so if you dont want to pass a max_health you might pass it

**Comment by DavidMistriel** [comment]
Im not sure what you mean. I check if the healing is more then mex health
if it is then set his hp to maxhealth if its not then add the healing

File: HeroGame/hero.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    def level_up(self):
+        if self.coins >= Hero.level_up_threshold * (self.level+1):
+            self.level += 1
+            self.damage += math.ceil(self.damage / 10)
+            self.max_health += math.ceil(self.damage / 10)
+            self.hp = self.max_health
+            self.coins -= Hero.level_up_threshold * self.level+1
```
Comment (alphabestteam) [inline HeroGame/hero.py:39]: the `+1` is after the `*` and is wrong calculation

File: HeroGame/hero.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    def attack(self, monster: object):
+        monster.reduce_health(self)
+
+    def reduce_health(self, monster: object) -> None:
+        if self.defending:
+            self.hp = self.hp - monster.damage * 0.20
+            self.defending = False
+            print("You dodged 80% of the damage")
+        else:
+            self.hp = self.hp - monster.damage
```
Comment (alphabestteam) [inline HeroGame/hero.py:50]: -=

File: HeroGame/hero.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    def hp(self, value):
+        self._hp = value
+
+    @property
+    def level(self):
+        return self._level
+
+    @level.setter
+    def level(self, value):
+        self._level = value
```
Comment (alphabestteam) [inline HeroGame/hero.py:101]: you are missing the function defend

File: HeroGame/main.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    hero.increase_coins(1)
+
+
+def choose_action(hero_name: str):
+    print(f"Its your turn {hero_name}, what would you like to do now?")
+    return input("1 -> Defend\n2 -> Attack\n3 -> Heal\n4 -> Level Up\n")
+
+
+def handle_monsters_turn(hero, monster):
+    monster.attack(hero)
```
Comment (alphabestteam) [inline HeroGame/main.py:39]: this does not need to be in a functions, it is a action of one line, and very detailed move from the name of the object and function

File: HeroGame/main.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    print("The monster attacked you.")
+
+
+def monster_is_dead(monster: Monster) -> bool:
```
Comment (alphabestteam) [inline HeroGame/main.py:43]: this does not need to be in a functions, it is a action of one line, and very detailed move from the name of the object and field

File: HeroGame/main.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        if monster_is_dead(monster):
+            print("You killed the enemy! You gain 10 coins.")
+            hero.increase_coins(10)
+            monster = Monster("FITZ", hero.level)
+
+        handle_monsters_turn(hero, monster)
+
+
+def end_dialog(hero: Hero):
+    print(f"You died at level {hero.level} with {hero.coins} coins left, good game!")
```
Comment (alphabestteam) [inline HeroGame/main.py:72]: this does not need to be in a functions, it is a action of one line called from one place....

**Pull request merged**